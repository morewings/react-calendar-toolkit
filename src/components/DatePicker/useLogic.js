import {useCallback} from 'react';
import {
  useDatePickerActions,
  useDatePickerContext,
  useIncrementPrecision,
} from 'features/datepicker';

const useLogic = (onDateSet, minPrecision) => {
  const {
    state: {precision},
  } = useDatePickerContext();
  const {setDate, setVisibility} = useDatePickerActions();
  const incrementPrecision = useIncrementPrecision(minPrecision);
  const handleDateSet = useCallback(
    date => {
      if (precision === minPrecision) {
        onDateSet(date);
        setDate(date);
      } else {
        setVisibility(date);
        incrementPrecision();
      }
    },
    [
      incrementPrecision,
      minPrecision,
      onDateSet,
      precision,
      setDate,
      setVisibility,
    ]
  );
  return {handleDateSet};
};

export default useLogic;
