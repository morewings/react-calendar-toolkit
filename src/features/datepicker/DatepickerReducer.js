import {SET_DATE, SET_TODAY, SET_PRECISION} from './actionTypes';

const initialState = {
  selectedTimestamp: 0,
  todayTimestamp: 0,
  precision: 'day',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_DATE: {
      const {selectedTimestamp, precision} = action.payload;
      return {
        ...state,
        selectedTimestamp,
        precision,
      };
    }
    case SET_TODAY: {
      const {todayTimestamp} = action.payload;
      return {
        ...state,
        todayTimestamp,
      };
    }
    case SET_PRECISION: {
      return {
        ...state,
        precision: action.payload,
      };
    }
    default:
      return state;
  }
};
