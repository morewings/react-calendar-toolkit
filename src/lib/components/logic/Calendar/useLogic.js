import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsSameMonth,
  checkIsWeekend,
  checkIsWithinInterval,
  floorMonth,
  matchDatesWithPrecision,
} from 'lib/utils/dateUtils';

const useLogic = ({
  disableDate,
  endDate,
  precision,
  startDate,
  highlightDate,
  onDateSet,
  highlightWeekends,
  visibleTimestamp,
  selectedTimestamp,
  todayTimestamp,
}) => {
  const getIsDisabled = useCallback(
    date => {
      if (precision === 'year') return false; // we are not disabling years, just rendering the range
      if (precision === 'month') {
        return !checkIsWithinInterval(
          {start: floorMonth(startDate), end: ceilMonth(endDate)},
          date
        );
      }
      return (
        disableDate({date, isWeekend: checkIsWeekend(date), precision}) ||
        !checkIsWithinInterval({start: startDate, end: endDate}, date)
      );
    },
    [disableDate, endDate, precision, startDate]
  );

  const getIsHighlighted = useCallback(
    date => highlightDate({date, isWeekend: checkIsWeekend(date), precision}),
    [highlightDate, precision]
  );

  const handleDateSet = useCallback(
    date => {
      const isDisabled = getIsDisabled(date);
      !isDisabled && onDateSet(date);
    },
    [getIsDisabled, onDateSet]
  );

  const getIsWeekend = date =>
    precision === 'day' && highlightWeekends && checkIsWeekend(date);

  const getBelongsToSameMonth = date =>
    precision === 'day' && checkIsSameMonth(date, visibleTimestamp);

  const getIsSelected = date =>
    matchDatesWithPrecision(precision, date, selectedTimestamp);

  const getIsCurrent = date =>
    matchDatesWithPrecision(precision, date, todayTimestamp);

  return {
    getIsDisabled,
    getIsHighlighted,
    handleDateSet,
    getIsWeekend,
    getBelongsToSameMonth,
    getIsSelected,
    getIsCurrent,
  };
};

export default useLogic;
