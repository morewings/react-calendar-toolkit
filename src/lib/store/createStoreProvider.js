import React, {useReducer} from 'react';

const createStoreProvider = ({initialState, reducer, context}) => ({
  // eslint-disable-next-line react/prop-types
  children,
}) => {
  const {Provider} = context;
  const [state, dispatch] = useReducer(reducer, initialState);
  return <Provider value={{state, dispatch}}>{children}</Provider>;
};

export default createStoreProvider;
