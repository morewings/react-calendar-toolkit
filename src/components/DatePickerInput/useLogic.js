import {useCallback} from 'react';
import {useModalActions} from 'features/modal';
import {convertToDate} from 'utils/dateUtils';
import {useFormatDate} from 'features/locale';
import {useDatePickerContext} from 'features/datepicker';

const useLogic = ({onDateSet, hideOnSelect}) => {
  const {toggleDatePicker} = useModalActions();
  const formatDate = useFormatDate();
  const {
    state: {selectedTimestamp},
  } = useDatePickerContext();

  const handleDateSet = useCallback(
    date => {
      onDateSet(date);
      hideOnSelect && toggleDatePicker(false);
    },
    [hideOnSelect, onDateSet, toggleDatePicker]
  );

  const formatValue = useCallback(
    formatPattern =>
      formatDate(formatPattern, convertToDate(selectedTimestamp)),
    [formatDate, selectedTimestamp]
  );

  const getDate = () => convertToDate(selectedTimestamp);

  return {handleDateSet, formatValue, getDate};
};

export default useLogic;
