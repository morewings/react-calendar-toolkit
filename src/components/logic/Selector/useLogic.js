import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsWithinInterval,
  decrementMonth,
  floorMonth,
  incrementMonth,
} from 'utils/dateUtils';
import {useDatePickerActions} from 'features/datepicker';

const useLogic = ({startDate, endDate, visibleTimestamp}) => {
  const {setVisibility, setPrecision} = useDatePickerActions();

  const isAvailable = useCallback(
    date =>
      checkIsWithinInterval(
        {start: floorMonth(startDate), end: ceilMonth(endDate)},
        date
      ),
    [endDate, startDate]
  );

  const onSetPrecision = useCallback(
    nextPrecision => {
      setPrecision(nextPrecision);
    },
    [setPrecision]
  );

  const onIncrementMonth = useCallback(() => {
    const nextDate = incrementMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  const onDecrementMonth = useCallback(() => {
    const nextDate = decrementMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  return {onSetPrecision, onIncrementMonth, onDecrementMonth};
};

export default useLogic;
