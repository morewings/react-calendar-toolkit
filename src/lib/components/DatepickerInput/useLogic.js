import {useCallback} from 'react';
import {useModalActions} from 'lib/features/modal';
import {convertToDate} from 'lib/utils/dateUtils';
import {useFormatDate} from 'lib/features/locale';
import {useDatepickerContext} from 'lib/features/datepicker';

const useLogic = ({onDateSet, hideOnSelect}) => {
  const {toggleDatepicker} = useModalActions();
  const formatDate = useFormatDate();
  const {
    state: {selectedTimestamp},
  } = useDatepickerContext();

  const handleDateSet = useCallback(
    date => {
      onDateSet(date);
      hideOnSelect && toggleDatepicker(false);
    },
    [hideOnSelect, onDateSet, toggleDatepicker]
  );

  const formatValue = useCallback(
    formatPattern =>
      formatDate(formatPattern, convertToDate(selectedTimestamp)),
    [formatDate, selectedTimestamp]
  );

  const getDate = () => convertToDate(selectedTimestamp);

  return {handleDateSet, formatValue, getDate};
};

export default useLogic;
