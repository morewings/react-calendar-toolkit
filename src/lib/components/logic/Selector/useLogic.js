import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsWithinInterval,
  subMonth,
  floorMonth,
  addMonth,
} from 'lib/utils/dateUtils';
import {useDatepickerActions} from 'lib/features/datepicker';

const useLogic = ({startDate, endDate, visibleTimestamp}) => {
  const {setVisibility, setPrecision} = useDatepickerActions();

  const isAvailable = useCallback(
    date =>
      checkIsWithinInterval(
        {start: floorMonth(startDate), end: ceilMonth(endDate)},
        date
      ),
    [endDate, startDate]
  );

  const onSetPrecision = useCallback(
    nextPrecision => {
      setPrecision(nextPrecision);
    },
    [setPrecision]
  );

  const onIncrementMonth = useCallback(() => {
    const nextDate = addMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  const onDecrementMonth = useCallback(() => {
    const nextDate = subMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  return {onSetPrecision, onIncrementMonth, onDecrementMonth};
};

export default useLogic;
