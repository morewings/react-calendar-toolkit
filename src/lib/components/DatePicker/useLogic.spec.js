import React from 'react';
import {renderHook, act} from '@testing-library/react-hooks';
import createStoreProvider from 'lib/store';
import DatePickerContext from 'lib/features/datepicker/context';
import DatepickerReducer from 'lib/features/datepicker/DatepickerReducer';
import useLogic from './useLogic';

const mockReducer = jest.fn((...args) => {
  DatepickerReducer(...args);
});

const onDateSet = jest.fn();

const date = new Date(2020, 0, 12); // 12.01.2020

const createStoreProviderWithState = state =>
  createStoreProvider({
    initialState: state,
    reducer: mockReducer,
    context: DatePickerContext,
  });

describe('DatePicker > useLogic > handleDateSet', () => {
  beforeEach(() => {
    onDateSet.mockClear();
    mockReducer.mockClear();
  });

  it('returns function', () => {
    const Provider = createStoreProviderWithState({});
    const {result} = renderHook(() => useLogic(onDateSet, 'day'), {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });
    expect(result.current.handleDateSet).toBeInstanceOf(Function);
  });

  it('sets date and calls callback when `precision === minPrecision`', () => {
    const precision = 'day';
    const Provider = createStoreProviderWithState({precision});
    const {result} = renderHook(() => useLogic(onDateSet, precision), {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });
    act(() => {
      result.current.handleDateSet(date);
    });
    expect(onDateSet).toHaveBeenCalledTimes(1);
    expect(onDateSet).toHaveBeenCalledWith(date);
    expect(mockReducer).toHaveBeenCalledTimes(1);
    expect(mockReducer.mock.calls[0][1]).toMatchSnapshot(); // SET_DATE
  });

  it('increments precision and sets visibility when `precision !== minPrecision`', () => {
    const minPrecision = 'day';
    const precision = 'month';
    const Provider = createStoreProviderWithState({precision});
    const {result} = renderHook(() => useLogic(onDateSet, minPrecision), {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });
    act(() => {
      result.current.handleDateSet(date);
    });
    expect(onDateSet).not.toHaveBeenCalled();
    expect(mockReducer).toHaveBeenCalledTimes(2);
    expect(mockReducer.mock.calls[0][1]).toMatchSnapshot(); // SET_VISIBILITY
    expect(mockReducer.mock.calls[1][1]).toMatchSnapshot(); // SET_PRECISION
  });
});
