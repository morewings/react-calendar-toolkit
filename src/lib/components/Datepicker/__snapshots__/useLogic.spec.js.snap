// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datepicker > useLogic > handleDateSet increments precision and sets visibility when \`precision !== minPrecision\` 1`] = `
Object {
  "payload": Object {
    "visibleTimestamp": 1578787200000,
  },
  "type": "SET_VISIBILITY",
}
`;

exports[`Datepicker > useLogic > handleDateSet increments precision and sets visibility when \`precision !== minPrecision\` 2`] = `
Object {
  "payload": "day",
  "type": "SET_PRECISION",
}
`;

exports[`Datepicker > useLogic > handleDateSet sets date and calls callback when \`precision === minPrecision\` 1`] = `
Object {
  "payload": Object {
    "selectedTimestamp": 1578787200000,
  },
  "type": "SET_DATE",
}
`;
